File: pipeline.cpp
#include <cmath>
#include <deque>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <string>
#include <unordered_map>

constexpr int ROLLING_WINDOW = 3;
constexpr double ANOMALY_THRESHOLD_PCT = 0.8;

struct SymbolState {
    long long total_volume = 0;
    double total_notional = 0.0;
    std::deque<double> rolling_prices;
    double rolling_mean_last = 0.0;
    double last_price = -1.0;
    double peak_price = -1.0;
    double max_drawdown_pct = 0.0;
    int anomaly_count = 0;

    void update(double price, int volume) {
        total_volume += volume;
        total_notional += price * volume;

        rolling_prices.push_back(price);
        if (static_cast<int>(rolling_prices.size()) > ROLLING_WINDOW) {
            rolling_prices.pop_front();
        }

        double rolling_sum = 0.0;
        for (double p : rolling_prices) {
            rolling_sum += p;
        }
        rolling_mean_last = rolling_sum / rolling_prices.size();

        if (last_price > 0.0) {
            double jump_pct = std::abs((price - last_price) / last_price) * 100.0;
            if (jump_pct >= ANOMALY_THRESHOLD_PCT) {
                anomaly_count++;
            }
        }

        if (price > peak_price) {
            peak_price = price;
        }
        if (peak_price > 0.0) {
            double drawdown_pct = ((peak_price - price) / peak_price) * 100.0;
            if (drawdown_pct > max_drawdown_pct) {
                max_drawdown_pct = drawdown_pct;
            }
        }

        last_price = price;
    }
};

bool parse_line(const std::string& line, std::string& symbol, double& price, int& volume) {
    std::stringstream ss(line);
    std::string timestamp;
    std::string price_str;
    std::string volume_str;

    if (!std::getline(ss, timestamp, ',')) return false;
    if (!std::getline(ss, symbol, ',')) return false;
    if (!std::getline(ss, price_str, ',')) return false;
    if (!std::getline(ss, volume_str, ',')) return false;

    try {
        price = std::stod(price_str);
        volume = std::stoi(volume_str);
    } catch (...) {
        return false;
    }
    return !symbol.empty();
}

int main(int argc, char* argv[]) {
    if (argc != 3) {
        std::cerr << "Usage: ./pipeline <input_csv> <output_json>\n";
        return 1;
    }

    std::ifstream input(argv[1]);
    if (!input.is_open()) {
        std::cerr << "Failed to open input file\n";
        return 1;
    }

    std::unordered_map<std::string, SymbolState> states;
    std::string line;

    std::getline(input, line); // header

    while (std::getline(input, line)) {
        std::string symbol;
        double price = 0.0;
        int volume = 0;
        if (!parse_line(line, symbol, price, volume)) {
            continue;
        }
        states[symbol].update(price, volume);
    }

    std::ofstream output(argv[2]);
    if (!output.is_open()) {
        std::cerr << "Failed to open output file\n";
        return 1;
    }

    output << "{\n";
    bool first_symbol = true;
    for (const auto& entry : states) {
        const auto& symbol = entry.first;
        const auto& state = entry.second;
        double vwap = state.total_volume > 0 ? state.total_notional / state.total_volume : 0.0;

        if (!first_symbol) {
            output << ",\n";
        }
        first_symbol = false;

        output << "  \"" << symbol << "\": {\n";
        output << "    \"total_volume\": " << state.total_volume << ",\n";
        output << std::fixed << std::setprecision(4);
        output << "    \"vwap\": " << vwap << ",\n";
        output << "    \"rolling_mean_last\": " << state.rolling_mean_last << ",\n";
        output << "    \"anomaly_count\": " << state.anomaly_count << ",\n";
        output << "    \"max_drawdown_pct\": " << state.max_drawdown_pct << "\n";
        output << "  }";
    }
    output << "\n}\n";

    std::cout << "Processed " << states.size() << " symbols. Output written to " << argv[2] << "\n";
    return 0;
}

File: pipeline.py
from __future__ import annotations

import csv
import json
import math
import sys
from collections import defaultdict, deque
from dataclasses import dataclass, field
from pathlib import Path
from typing import Deque, Dict, Iterable, List


ROLLING_WINDOW = 3
ANOMALY_THRESHOLD_PCT = 0.8


@dataclass
class SymbolState:
    total_volume: int = 0
    total_notional: float = 0.0
    rolling_prices: Deque[float] = field(default_factory=lambda: deque(maxlen=ROLLING_WINDOW))
    rolling_mean_last: float = 0.0
    last_price: float | None = None
    peak_price: float = -math.inf
    max_drawdown_pct: float = 0.0
    anomaly_count: int = 0

    def update(self, price: float, volume: int) -> None:
        self.total_volume += volume
        self.total_notional += price * volume

        self.rolling_prices.append(price)
        self.rolling_mean_last = round(sum(self.rolling_prices) / len(self.rolling_prices), 4)

        if self.last_price is not None and self.last_price != 0:
            price_jump_pct = abs((price - self.last_price) / self.last_price) * 100
            if price_jump_pct >= ANOMALY_THRESHOLD_PCT:
                self.anomaly_count += 1

        self.peak_price = max(self.peak_price, price)
        if self.peak_price > 0:
            drawdown_pct = ((self.peak_price - price) / self.peak_price) * 100
            self.max_drawdown_pct = max(self.max_drawdown_pct, drawdown_pct)

        self.last_price = price

    def to_dict(self) -> Dict[str, float | int]:
        vwap = round(self.total_notional / self.total_volume, 4) if self.total_volume else 0.0
        return {
            "total_volume": self.total_volume,
            "vwap": round(vwap, 4),
            "rolling_mean_last": round(self.rolling_mean_last, 4),
            "anomaly_count": self.anomaly_count,
            "max_drawdown_pct": round(self.max_drawdown_pct, 4),
        }


def parse_rows(csv_path: Path) -> Iterable[dict]:
    with csv_path.open("r", newline="", encoding="utf-8") as handle:
        reader = csv.DictReader(handle)
        for row in reader:
            if not row.get("timestamp") or not row.get("symbol"):
                continue
            try:
                row["price"] = float(row["price"])
                row["volume"] = int(row["volume"])
            except (TypeError, ValueError):
                continue
            yield row


def process_ticks(csv_path: Path) -> Dict[str, Dict[str, float | int]]:
    states: Dict[str, SymbolState] = defaultdict(SymbolState)
    for row in parse_rows(csv_path):
        states[row["symbol"]].update(price=row["price"], volume=row["volume"])
    return {symbol: state.to_dict() for symbol, state in sorted(states.items())}


def main(argv: List[str]) -> int:
    if len(argv) != 3:
        print("Usage: python3 pipeline.py <input_csv> <output_json>")
        return 1

    input_path = Path(argv[1])
    output_path = Path(argv[2])

    results = process_ticks(input_path)
    output_path.write_text(json.dumps(results, indent=2), encoding="utf-8")
    print(f"Processed {len(results)} symbols. Output written to {output_path}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv))
